yapapi/contrib/service/http_proxy.py

```diff
@@ -0,0 +1,173 @@
+import abc
+import aiohttp
+from aiohttp import web, client_ws
+import asyncio
+import logging
+from multidict import CIMultiDict
+import re
+from typing import Optional
+import traceback
+
+from yapapi.services import Cluster, ServiceState, Service
+
+
+logger = logging.getLogger(__name__)
+
+
+class _ResponseParser:
+    def __init__(self, ws: client_ws.ClientWebSocketResponse, timeout: float = 10.0):
+        self.headers_data = b""
+        self.headers: Optional[CIMultiDict] = None
+        self.content = b""
+        self.status: Optional[int] = None
+        self.ws = ws
+        self.timeout = timeout
+
+    def process_headers(self):
+        header_lines = self.headers_data.decode("ascii").splitlines()
+
+        self.status = int(header_lines[0].split()[1])
+        self.headers = CIMultiDict()
+
+        for line in header_lines[1:]:
+            if line:
+                name, value = line.split(": ", maxsplit=1)
+                self.headers[name] = value
+
+    def receive_data(self, data: bytes):
+        if self.status:
+            self.content += data
+        else:
+            parts = re.split(b"\r\n\r\n", data)
+            self.headers_data += parts.pop(0)
+            if parts:
```

Comment:
Do I understand correctly that headers always end with `b"\r\n\r\n"`?
Because if not, then empty content will hang forever.

Reply:
yes, according to the RFC, the headers always end on an empty line ending with CRLF (and the previous CRLF belongs to the either the status line or to the last line of the headers)

ttps://www.w3.org/Protocols/rfc2616/rfc2616-sec4.html